app/Actions/Album/PositionData.php

```diff
@@ -45,6 +45,9 @@ public function get(AbstractAlbum $album, bool $includeSubAlbums = false): array
 			->whereNotNull('longitude')
 			->get()
 			->toArray();
+		if ($album instanceof Album) {
+			$result['track_short_path'] = $album->track_short_path;
+		}
```

Comment:
Much better :+1:

```suggestion
		$result['track_url'] = $album instanceof Album ? $album->track_url : null;
```

But I have to excuse, that I didn't told you the whole truth earlier. My previous explanation was incomplete. 

In the DB we store the _short path_ which is relative to a certain disk of the Flysystem File adapter. But we need to send a _URL_ to the browser. The Flysystem storage provides the method `url(string $shortPath)` for that. The will become crucial, when some day in the future the disk isn't the same (local) disk on which Lychee is installed, but for example a S3 bucket.

The whole aspect should be in the same spirit as already done for `SizeVariant` or `Symlink`. For example the model `SizeVariant` provides to attributes: `short_path` and `url` (Please ignore the `full_path`. This is evil and will break with other than local storage. It is only there for legacy reasons, because our image manipulation code is based on absolute paths and does not work with proper file streams.)

https://github.com/LycheeOrg/Lychee/blob/019921260b4d8e1bac0e61ec99fd4d8782c68710/app/Models/SizeVariant.php#L40-L48

The `Album` model should have two attributes as well `track_short_path` and `track_url`.

The attribute `track_url` shall be a read-only attribute. To enforce that, the `Album` model shall use the `MustNotSet` cast. As a template see again `SizeVariant`:

https://github.com/LycheeOrg/Lychee/blob/019921260b4d8e1bac0e61ec99fd4d8782c68710/app/Models/SizeVariant.php#L70-L78

Moreover, the `Album` model needs a getter which provides the attribute `track_url` based on `track_short_path`. The role model is

https://github.com/LycheeOrg/Lychee/blob/019921260b4d8e1bac0e61ec99fd4d8782c68710/app/Models/SizeVariant.php#L123-L136

Please forget all the complexity of the method above. In your case the additional getter method on the `Album` model will be rather simple:

```php
	/**
	 * Accessor for the "virtual" attribute {@link SizeVariant::$url}.
	 *
	 * This is convenient method which wraps
	 * {@link Album::$track_short_path} into
	 * {@link \Illuminate\Support\Facades\Storage::url()}.
	 *
	 * @return string the url of the track
	 */
	public function getTrackUrlAttribute(): string
	{
			return Storage::url($this->track_short_path);
	}
```

A final word about the ternary operator. The returned JSON object should always be identical. (This is necessary for statically, strongly typed clients.) Hence, if we have a tag album, the attribute `track_url` should still be provided, but be `null`.

Reply:
I don't work with PHP usually, so what would be the "correct" way to add the `track_url` attribute with value `null` to tag albums (and smart albums, I guess, should have this as well)?

Reply:
See the last line of my suggestion above. Here again:

```ph$result['track_url'] = $album instanceof Album ? $album->track_url : null;
```

Reply:
This is only used in `PosutionData` - what hapoens if directly requestijg Albums (`Abum::get` and `Albuns::get`)?

Reply:
Ah, I see. Then we had a misunderstanding here. I didn't want to say that the attribute must be everywhere, I only wanted to say that an object of  the same type (here: `PositionData`) should always have the same attributes. So it is fine if `TagAlbum` and `SmartAlbum` don't return the attribute.

However, speaking about the serialization of `Album`, I noted another aspect which needs attention. When converted to JSON a `Album` object should include the "virtual" attribute `track_url`, but skip the attribute `track_short_path`. The `Album` class contains two arrays for that: `append` and `hidden`. The name of the attributes must be added to these arrays accordingly.